torch/package/package_importer.py

```diff
@@ -337,20 +393,56 @@ def _make_module(
 
         return module
 
+    def _check_if_viable_parent(self, cur, child, name):
+        if isinstance(cur, _SelectiveExternNode):
+            return
+        if not isinstance(cur, (_PackageNode)) or child not in cur.children:
+            raise ModuleNotFoundError(
+                f'No module named "{name}" with in self-contained archive "{self.filename}"'
+                f" and the module is also not in the list of allowed external modules: {self.extern_modules}",
+                name=name,
+            )
+
+    def _get_node_from_name(self, name: str):
+        cur: _PathNode = self.root
+        atoms = name.split(".")
+        for atom in atoms[:-1]:
+            if isinstance(cur, _ExternNode):
+                return cur
+            if isinstance(cur, (_SelectiveExternNode, _PackageNode)):
+                self._check_if_viable_parent(cur, atom, name)
+                cur = cur.children[atom]
+        if isinstance(cur, (_SelectiveExternNode, _PackageNode)):
+            self._check_if_viable_parent(cur, atoms[-1], name)
+            return cur.children[atoms[-1]]
+        elif isinstance(cur, _ExternNode):
+            return cur
+
+
+
     def _load_module(self, name: str, parent: str):
         cur: _PathNode = self.root
         for atom in name.split("."):
-            if not isinstance(cur, _PackageNode) or atom not in cur.children:
-                raise ModuleNotFoundError(
-                    f'No module named "{name}" in self-contained archive "{self.filename}"'
-                    f" and the module is also not in the list of allowed external modules: {self.extern_modules}",
-                    name=name,
-                )
-            cur = cur.children[atom]
+            self._check_if_viable_parent(cur, atom, name)
+            if isinstance(cur, _SelectiveExternNode):
+                if atom not in cur.children:
+                    module = self.modules[name] = importlib.import_module(name)
+                    return module
+
+            if isinstance(cur, (_PackageNode, _SelectiveExternNode)):
```

Comment:
it's confusing to me that we have important logic both *before* and *after* L433, so that within the same loop body `cur` refers to both the parent and child. Can we refactor this so that the this `cur = cur.children[atom]` is set either at the beginning or the end?